### PR TITLE
net: coap: migrate to standard buffers

### PR DESCRIPTION
**Pull request by mike-scott**
In order to make the CoAP library more useful in a generic way, this patch series remove all references to net_buf and net_pkt and instead allow usage of standard char-based buffers which will hold generated packet data or be passed for parsing.

Highlights:
- Works with any networking backend in that it's merely a parsing / packet building library
- Supports standard buffers needed for sockets
- Doesn't make a copy of the existing library
- Drops about 800 lines of code in Zephyr

### CODE REVIEW

**Comment by mike-scott** [comment]
NOTE: This does not include the changes for LwM2M.

**Comment by mike-scott** [comment]
@rveerama1 I know you are also working on this and I really hate that we are duplicating, but this was a blocker for the work on LwM2M so I went forward a bit.

I'm very partial to this approach for many reasons, but I think if you look at the tests/net/lib/coap/src/main.c changes it really highlights the "why" I did it this way.

Please consider, thanks.

**Comment by mike-scott** [comment]
Also, this isn't really WIP code.  It has been actually tested a bit.  The tests/net/lib/coap passes, and the coap_client / coap_server are acting normally as far as I can tell.   I don't know that I have the right setup to test rpl_* and leds_demo samples, but they compile.

I'm sure sanity will find some nice issues for me tho.

**Comment by mike-scott** [comment]
@pfalcon the "flat buffer handling" change is purely a design choice and could be changed if you feel that it doesn't add value.   The code itself can also be relocated into an actual .c file somewhere else behind a Kconfig which can be selected by CoAP and LwM2M.

File: tests/net/lib/coap/src/main.c [hunk 1]
```diff
@@ -30,6 +30,11 @@
 #define NUM_OBSERVERS 3
 #define NUM_REPLIES 3
 
+#define MAX_PACKET_SIZE 256
+
+static u8_t test_buf[MAX_PACKET_SIZE];
+static u8_t test_buf2[MAX_PACKET_SIZE];
+
 NET_PKT_TX_SLAB_DEFINE(coap_pkt_slab, 4);
```
Comment (mike-scott) [inline tests/net/lib/coap/src/main.c]: Woops forgot to remove this stuff as well.

**Comment by rveerama1** [comment]
> @rveerama1 I know you are also working on this and I really hate that we are duplicating, but this was a blocker for the work on LwM2M so I went forward a bit.
> 
> I'm very partial to this approach for many reasons, but I think if you look at the tests/net/lib/coap/src/main.c changes it really highlights the "why" I did it this way.
> 
> Please consider, thanks.

@mike-scott I just created my WIP PR https://github.com/zephyrproject-rtos/zephyr/pull/10121. Your first version was completely different. This version looks similar to my changes except "flat_buf_util". I am not against about that, but these helpers will become soon de-facto standard for socket based protocols and sample applications.
There is a reason for creating separate folder for CoAP changes. We can deprecate current CoAP over net_pkt in this release and make sure every user modified according to updated CoAP library and then we can safely remove net_pkt based after release. Same thing happened when ZOAP converted to CoAP (+support for multi-fragments changes).
I am also adding extra functionality to coap_client to verify coap_server functionality. Current coap_client is very simple. That's why I am taking some time.

**Pull request closed without merging**